File: python/712.py
"""Project Euler Problem 712: Exponent difference."""

from math import isqrt

def prime_count_table(N):
    """
    Lucy_Hedgehog algorithm to compute pi(n/k) for all distinct quotients.
    Returns a function get(v) that gives pi(v) for any v that is either <= sqrt(N) or N//k for some k.
    """
    r = isqrt(N)
    # V contains all distinct values of N // i
    V = [N // i for i in range(1, r + 1)]
    V += list(range(V[-1] - 1, 0, -1))

    # S[v] = count of integers 2..v (initially v - 1)
    S = {v: v - 1 for v in V}

    for p in range(2, r + 1):
        if S[p] > S[p - 1]:  # p is prime
            sp = S[p - 1]
            p2 = p * p
            for v in V:
                if v < p2:
                    break
                S[v] -= S[v // p] - sp

    return S


def solve():
    N = 10**12
    M = 10**9 + 7
    L = isqrt(N)

    # Sieve primes up to N/L ~ sqrt(N)
    limit = N // L + 1
    sieve = [True] * (limit + 1)
    sieve[0] = sieve[1] = False
    for i in range(2, isqrt(limit) + 1):
        if sieve[i]:
            for j in range(i * i, limit + 1, i):
                sieve[j] = False
    primes_list = [i for i in range(2, limit + 1) if sieve[i]]

    ans = 0

    # Process small primes (p <= N/L ~ sqrt(N))
    for p in primes_list:
        counts = []
        e = 0
        pe = 1
        while pe <= N:
            pe_next = pe * p
            cnt = (N // pe - N // pe_next) % M
            counts.append(cnt)
            pe = pe_next
            e += 1

        # Sum over all pairs (vn, vm)
        for vn in range(len(counts)):
            for vm in range(len(counts)):
                diff = abs(vn - vm)
                contribution = diff * counts[vn] % M * counts[vm] % M
                ans = (ans + contribution) % M

    # Process large primes using prime counting
    # For large p (p > sqrt(N)), v_p(n) <= 1, so only contributions are
    # when v_p(n) = 1, v_p(m) = 0 or vice versa.
    # Number of n with v_p(n) = 0: N - N//p
    # Number of n with v_p(n) = 1: N//p - N//p^2 ~ N//p for large p
    # Contribution: 2 * (N - N//p) * (N//p) per prime

    # Sum over primes p where N//p takes value q
    # Number of such primes = pi(N/q) - pi(N/(q+1))
    S = prime_count_table(N)

    # We need to exclude small primes we already processed
    num_small_primes = len(primes_list)

    for q in range(1, L):
        # pi(N/q) counts primes p with p <= N/q, i.e., N/p >= q, i.e., floor(N/p) >= q
        # pi(N/(q+1)) counts primes p with p <= N/(q+1)
        # Difference = primes with q <= floor(N/p) < q+1, i.e., floor(N/p) = q
        pi_q = S.get(N // q, 0) if N // q > 0 else 0
        pi_q_plus_1 = S.get(N // (q + 1), 0) if N // (q + 1) > 0 else 0

        # But we need primes > limit (already handled small primes)
        if N // q <= limit:
            pi_q = min(pi_q, num_small_primes)
        if N // (q + 1) <= limit:
            pi_q_plus_1 = min(pi_q_plus_1, num_small_primes)

        num_primes_in_range = (pi_q - pi_q_plus_1) % M

        # For each prime p with floor(N/p) = q:
        # v_p(n) = 0 for N - q numbers, v_p(n) >= 1 for q numbers
        # Actually for large p, v_p(n) = 1 for exactly q numbers
        # Contribution: 2 * (N - q) * q per prime
        contribution = 2 * (N - q) % M * q % M * num_primes_in_range % M
        ans = (ans + contribution) % M

    return ans


if __name__ == "__main__":
    print(solve())

File: c/712.c
/*
 * Project Euler Problem 712: Exponent Difference.
 *
 * For each prime p, sum |v_p(n) - v_p(m)| over all 1 <= n, m <= N.
 * Small primes: enumerate all exponent counts directly.
 * Large primes (p > sqrt(N)): v_p(n) is 0 or 1, contribution = 2*(N - N/p)*(N/p).
 * Use Lucy DP for prime counting.
 */
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

typedef long long ll;
typedef __int128 lll;

#define MOD 1000000007LL

int main() {
    ll N = 1000000000000LL;
    ll L = (ll)sqrt((double)N);
    while ((L + 1) * (L + 1) <= N) L++;
    while (L * L > N) L--;

    int limit = (int)(N / L) + 1;

    /* Sieve primes up to limit */
    char *sieve = calloc(limit + 1, 1);
    for (int i = 2; i * i <= limit; i++) {
        if (!sieve[i]) {
            for (int j = i * i; j <= limit; j += i)
                sieve[j] = 1;
        }
    }
    int *primes_list = malloc((limit + 1) * sizeof(int));
    int num_small_primes = 0;
    for (int i = 2; i <= limit; i++) {
        if (!sieve[i]) primes_list[num_small_primes++] = i;
    }
    free(sieve);

    ll ans = 0;

    /* Process small primes */
    for (int pi = 0; pi < num_small_primes; pi++) {
        ll p = primes_list[pi];
        ll counts[60];
        int num_e = 0;
        ll pe = 1;
        while (pe <= N) {
            ll pe_next;
            if (pe > N / p) pe_next = N + 1; else pe_next = pe * p;
            ll cnt = ((N / pe) - (pe_next <= N ? N / pe_next : 0)) % MOD;
            counts[num_e++] = cnt;
            if (pe > N / p) break;
            pe *= p;
        }

        for (int vn = 0; vn < num_e; vn++) {
            for (int vm = 0; vm < num_e; vm++) {
                int diff = vn > vm ? vn - vm : vm - vn;
                ll contribution = (ll)diff % MOD * (counts[vn] % MOD) % MOD * (counts[vm] % MOD) % MOD;
                ans = (ans + contribution) % MOD;
            }
        }
    }

    /* Lucy DP for prime counting */
    int r = (int)L;
    int vlen = 2 * r + 2;
    ll *V = malloc(vlen * sizeof(ll));
    ll *S = malloc(vlen * sizeof(ll));
    /* Map: for v > r, index = N/v; for v <= r, index = r + v (approximately) */
    /* Actually use direct mapping via sorted array */

    int vcnt = 0;
    for (ll i = 1; i <= r; i++) {
        V[vcnt++] = N / i;
    }
    ll last = V[vcnt - 1];
    for (ll v = last - 1; v >= 1; v--) {
        V[vcnt++] = v;
    }

    /* S[i] = number of "primes" (actually integers >= 2) up to V[i], initially V[i] - 1 */
    /* We need a way to look up S[v]. For v > r, index = N/v - 1 (since V[0] = N/1, V[1] = N/2, etc.)
       For v <= r, index = vcnt - v. Actually let's use a function. */

    /* Better: use two arrays. small[v] for v <= r, big[i] for N/i, i <= r */
    ll *small_s = calloc(r + 2, sizeof(ll));
    ll *big_s = calloc(r + 2, sizeof(ll));

    for (int v = 1; v <= r; v++) small_s[v] = v - 1;
    for (int i = 1; i <= r; i++) big_s[i] = N / i - 1;

    for (int p = 2; p <= r; p++) {
        if (small_s[p] == small_s[p - 1]) continue; /* p is not prime */
        ll sp = small_s[p - 1];
        ll p2 = (ll)p * p;
        for (int i = 1; i <= r && N / i >= p2; i++) {
            ll v_over_p = N / i / p;
            ll sv;
            if (v_over_p <= r) sv = small_s[v_over_p];
            else sv = big_s[i * p]; /* N/(i*p) is the index */
            /* Wait, we need S[N/(i*p)]. If i*p <= r, that's big_s[i*p].
               Otherwise if N/(i*p) <= r, it's small_s[N/(i*p)]. */
            if ((ll)i * p <= r) sv = big_s[i * p];
            else sv = small_s[(int)(N / ((ll)i * p))];
            big_s[i] -= sv - sp;
        }
        for (int v = r; v >= p2; v--) {
            small_s[v] -= small_s[v / p] - sp;
        }
    }

    /* Now small_s[v] = pi(v) for v <= r, big_s[i] = pi(N/i) for i <= r */
    /* We need pi(N/q) and pi(N/(q+1)) for q = 1..L-1 */

    for (int q = 1; q < (int)L; q++) {
        /* pi(N/q) */
        ll pi_q;
        ll nq = N / q;
        if (nq <= r) pi_q = small_s[(int)nq]; else pi_q = big_s[q];

        ll pi_q1;
        ll nq1 = N / (q + 1);
        if (nq1 <= r) pi_q1 = small_s[(int)nq1]; else pi_q1 = big_s[q + 1];

        /* Exclude small primes we already processed */
        if (nq <= limit) {
            if (pi_q > num_small_primes) pi_q = num_small_primes;
        }
        if (nq1 <= limit) {
            if (pi_q1 > num_small_primes) pi_q1 = num_small_primes;
        }

        ll num_primes_in_range = (pi_q - pi_q1) % MOD;
        if (num_primes_in_range < 0) num_primes_in_range += MOD;

        /* Contribution: 2 * (N - q) * q * num_primes_in_range */
        ll contribution = 2 * ((N - q) % MOD) % MOD * (q % MOD) % MOD * num_primes_in_range % MOD;
        ans = (ans + contribution) % MOD;
    }

    printf("%lld\n", ans);

    free(primes_list);
    free(small_s);
    free(big_s);
    free(V);
    free(S);
    return 0;
}
